src/orchestrator-kubernetes/src/lib.rs

```diff
@@ -204,6 +250,78 @@ impl NamespacedKubernetesOrchestrator {
+        };
+        /// Get metrics for a particular service and process, converting them into a sane (i.e., numeric) format.
+        ///
+        /// Note that we want to keep going even if a lookup fails for whatever reason,
+        /// so this function is infallible. If we fail to get cpu or memory for a particular pod,
+        /// we just log a warning and install `None` in the returned struct.
+        async fn get_metrics(
+            self_: &NamespacedKubernetesOrchestrator,
+            id: &str,
+            i: i32,
+        ) -> ServiceProcessMetrics {
+            let name = format!("{id}-{i}");
+            let metrics = match self_.metrics_api.get(&name).await {
+                Ok(metrics) => metrics,
+                Err(e) => {
+                    warn!("Failed to get metrics for {name}: {e}");
+                    return ServiceProcessMetrics::default();
+                }
+            };
+            let Some(PodMetricsContainer { usage: PodMetricsContainerUsage { cpu: Quantity(cpu_str), memory: Quantity(mem_str) }, .. }) = metrics.containers.get(0) else {
+                warn!("metrics result contained no containers for {name}");
+                return ServiceProcessMetrics::default();
+            };
+            // Parsing a k8s Quantity is kind of complicated:
+            // See https://github.com/kubernetes/apimachinery/blob/master/pkg/api/resource/quantity.go
+            // for details. We can rewrite this correct parsing code in Rust if needed in the future, but let's see if we can get away with
+            // just assuming CPU is always a positive integer number of nano-CPUs and memory is always in KiB for now.
+            //
+            // Note that https://github.com/sombralibre/k8s-quantity-parser doesn't support the (undocumented) `n` suffix,
+            // so we can't use it.
+            let cpu = if cpu_str.ends_with('n') {
+                let rest = &cpu_str[0..cpu_str.len() - 1];
+                match rest.parse::<u64>() {
+                    Ok(val) => Some(val),
+                    Err(_e) => {
+                        tracing::warn!("metrics-server CPU value in unexpected format: {cpu_str}");
+                        None
+                    }
+                }
+            } else {
+                tracing::warn!("metrics-server CPU value in unexpected format: {cpu_str}");
+                None
+            };
+            let memory = if mem_str.ends_with("Ki") {
+                let rest = &cpu_str[0..mem_str.len() - 2];
+                match rest.parse::<u64>() {
+                    Ok(val) => Some(val),
+                    Err(_e) => {
+                        tracing::warn!(
+                            "metrics-server memory value in unexpected format: {mem_str}"
+                        );
+                        None
+                    }
+                }
+            } else {
+                tracing::error!("metrics-server memory value in unexpected format: {mem_str}");
```

Comment:
Should this be `error!` when all the rest are `warn!`?

Reply:
`error` was intentional, since we would want to know relatively urgently if the format somehow changes, since that would mean this code is globally wrong, not just that there's an issue with one environment.

(I even considered making it `panic!`)

Reply:
I will make the other ones `error`; making them `warn` was accidental